src/compiler/checker.ts

```diff
@@ -18196,6 +18196,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (objectType.flags & (TypeFlags.Any | TypeFlags.Never)) {
                 return objectType;
             }
+            if (indexType === neverType || indexType === implicitNeverType) {

```

Comment:
I'd probably go with this:
```suggestion
            if (indexType.flags & TypeFlags.Never && indexType !== silentNeverType) {
```

Reply:
In that way we would change the indexing behaviour of `unreachableNeverType` too. Are you sure? XD

Reply:
Am I sure? Not at all. It seems reasonable to treat that the same way as the regular never here but who knows?

Reply:
@Andarist 

I tried the following:
```ts
const obj = {
  arr: [],
};

const objWithIndex: Record<string, boolean> = {};

function testUnreachableNeverType(smt: number | string) {
  if(typeof smt === "number") {}
  else if(typeof smt === "string") {}
  else {
    const result = objWithIndex[smt]
  }
}
```

My supposition was that the type of `smt` is `unreachableNeverType`, therefore the type of `result` should have been `boolean` without handling `unreachableNeverType` in that `if` condition, but it is already `never` 🧐

I probably have to come up with a better example to try to have an inconsistency with `unreachableNeverType` too.